haskell/asterius/defs.bzl

```
@@ -31,3 +38,411 @@ asterius_toolchain = rule(
+            default_runfiles = ctx.attr.dep[DefaultInfo].default_runfiles,
+        ),
+    ]
+
+_asterius_untar = rule(
+    _asterius_untar_impl,
+    attrs = {
+        "dep": attr.label(
+            mandatory = True,
+            allow_single_file = True,
+            doc = "The ahc_dist target producing the archive.",
+        ),
+        "subfolder_name": attr.string(
+            doc = "The name of the new folder where the archive will be extracted",
+        ),
+    },
+    toolchains = [
+        "@rules_sh//sh/posix:toolchain_type",
+    ],
+    doc = "extract the files from the archive produced by ahc_dist",
+)
+
+def asterius_webpack_impl(ctx):
+    ahc_dist_info = ctx.attr.ahc_dist_dep[AhcDistInfo]
+    if ahc_dist_info.target != "browser":
+        fail("{} was built with target attribute '{}' but rule asterius_webpack_impl only works with the 'browser' target.".format(
+            ctx.attr.ahc_dist_dep.label,
+            ahc_dist_info.target,
+        ))
+
+    output_file = ctx.actions.declare_file("{}.mjs".format(ctx.label.name))
+    posix_toolchain = ctx.toolchains["@rules_sh//sh/posix:toolchain_type"]
+    all_paths = posix_toolchain.paths + [paths.dirname(ctx.executable._webpack.path)]
+
+    webpack_config_path = ahc_dist_info.webpack_config.path
+    webpack_command = " ".join([
+        "webpack.sh",
+        "--nobazel_node_patches",  # https://github.com/bazelbuild/rules_nodejs/issues/2076
+        "--config",
+        webpack_config_path,
+        "-o .",
+        "--output-filename",
+        output_file.path,
+    ])
+
+    ctx.actions.run_shell(
+        inputs = ctx.files.ahc_dist_dep,
+        outputs = [output_file],
+        command = " && ".join([
+            "tar xf {} --directory {}".format(ctx.file.ahc_dist_dep.path, paths.dirname(ctx.file.ahc_dist_dep.path)),
+            webpack_command,
+        ]),
+        env = {"PATH": ":".join(all_paths)},
+        arguments = [],
+        tools = ctx.files._webpack,
+    )
+
+    return [DefaultInfo(files = depset([output_file]))]
+
+asterius_webpack = rule(
+    asterius_webpack_impl,
+    attrs = {
+        "ahc_dist_dep": attr.label(
+            mandatory = True,
+            allow_single_file = True,
+            doc = """
+The ahc_dist target (built with target="browser") from which we will create a bundle.
+                  """,
+        ),
+        "_webpack": attr.label(
+            default = "@rules_haskell//haskell/asterius:webpack",
+            executable = True,
+            cfg = "exec",
+        ),
+    },
+    toolchains = [
+        "@rules_sh//sh/posix:toolchain_type",
+    ],
+    doc = "Creates a bundle using webpack out of ahc_dist target build for the browser.",
+)
+
+# Copied from https://github.com/tweag/asterius/blob/9f2574d9c2b50aa83d105741799e2f65b05e2023/asterius/test/ghc-testsuite.hs
+# The node options required to execute outputs from asterius.
+
+node_options = [
```

```suggestion
_node_options = [
```
Do these need to be public?